Feat/121 implement getavailablefetcherapis call

luca-schlecker opened the pull request:
Closes #121 

## What I have made

Forward `IFetcherService.getAvailableFetchers()` to the user.

## Checklist

Either tick or cross out the items that do not apply (using \~\~example text\~\~) and give a reason why the item does
not apply.

- [x] I have updated the documentation accordingly and commented my code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] (for reviewer) I have checked the implementation against the requirements

luca-schlecker committed:
feat: add IFetcherService interface and FetcherService implementation

mowi12 commented on src/main/kotlin/se/uulm/snowballr/backend/Module.kt:
Please add a new bullet point at the correction location in the documentation for the `snowballRModule` for the newly added fetcher service.

But is this even needed? We do not inject the service anywhere, so why is this needed.

luca-schlecker replied:
Not technically needed, I'll remove it.

mowi12 commented on src/main/kotlin/se/uulm/snowballr/backend/grpc/SnowballRServer.kt:
I thought about this implementation a while, but am not quite sure if this is the best solution.
Why don't we return the built object inside the service call implementation itself?

As I understand it, the `IFetcherService` is the interface which all the plugin fetchers need to implement, correct?
If yes, then this implementation is fine, else I don't see the reason not return `Main.AvailableFetcherApis` directly from the `getAvailableFetchers()` implementation in the service instead of set of strings.

luca-schlecker replied:
As an `IFetcherService` will later be used to give to a python fetcher for it to access other fetchers, the `Set<String>` is the best option.

mowi12 commented on src/main/kotlin/se/uulm/snowballr/backend/service/MainService.kt:
```kotlin
/**
 * The [MainService] class serves as the primary service implementation layer that aggregates multiple sub-services.
 * This class implements the [IMainService] interface and delegates the execution of specific functionality to the
 * sub-service implementations, e.g. [ProjectService].
 *
 * @constructor Initializes the [MainService] with the required repositories.
 * @param projectRepo The repository responsible for handling persistence operations related to projects.
 * @param criterionRepo The repository responsible for handling persistence operations related to criteria.
 * @param userRepo The repository responsible for handling persistence operations related to users.
 * @param projectMemberRepo The repository responsible for handling persistence operations related to project members.
 * @param jwtService The utility for handling JWT operations, such as token parsing and validation.
 */
class MainService(
    private val projectRepo: IProjectTableRepo,
    private val criterionRepo: ICriterionTableRepo,
    private val userRepo: IUserTableRepo,
    private val projectMemberRepo: IProjectMemberTableRepo,
    private val jwtService: IJwtService,
) : IMainService,
    IProjectService by ProjectService(projectRepo, userRepo),
    ICriterionService by CriterionService(criterionRepo),
    IUserService by UserService(userRepo, projectMemberRepo, jwtService),
    IFetcherService by FetcherService()
```

I think this is the correct instantiation of the fetcher service.

mowi12 commented on src/test/kotlin/se/uulm/snowballr/backend/service/MainServiceTest.kt:
These changes are not needed any more with the new instantiation.

mowi12 commented on src/main/kotlin/se/uulm/snowballr/backend/service/FetcherService.kt:
Please complete your documentation. Add all return types and parameters, as this is an important part of the codebase. As well as, format your documentation correctly. There should be a space between the description and the `@params` list.

mowi12 commented on src/main/kotlin/se/uulm/snowballr/backend/service/FetcherService.kt:
```diff
@@ -0,0 +1,134 @@
+    /**
+     * Get a list of the names of all registered fetchers.
+     */
+    suspend fun getAvailableFetchers(): Set<String>
```
I already added a comment in the `SnowballRServer` file about this.
As I now understood your code a bit more, I think this function should not return the set of string, but the already built object, that will be returned to the user.

This way we extract all the service logic into the service layer.

I know, there are currently implementations in the service layer, such as the register call, which also has some logic there, but this is already refactored and is currently in review.

luca-schlecker replied:
See [this comment](https://github.com/SE-UUlm/snowballr-backend/pull/126/files#r2208928742) for the reasoning behind this decision.

mowi12 commented on src/main/kotlin/se/uulm/snowballr/backend/service/FetcherService.kt:
```diff
@@ -0,0 +1,134 @@
+    /**
+     * Search for papers using a search query.
+     *
+     * @param fetcher The name of the fether to be used for this request.
```
```suggestion
     * @param fetcher The name of the fetcher to be used for this request.
```

luca-schlecker replied:
Already fixed.

mowi12 commented on src/main/kotlin/se/uulm/snowballr/backend/service/FetcherService.kt:
```diff
@@ -0,0 +1,134 @@
+     *    Paper A <-referring-- Paper B <-referring-- Paper C
+     *  forward ref.                                backward ref.
+     *
+     * @param fetcher The name of the fether to be used for this request.
```
```suggestion
     * @param fetcher The name of the fetcher to be used for this request.
```

luca-schlecker replied:
Already fixed.

mowi12 commented on src/main/kotlin/se/uulm/snowballr/backend/service/FetcherService.kt:
```diff
@@ -0,0 +1,134 @@
+     *    Paper A <-referring-- Paper B <-referring-- Paper C
+     *  forward ref.                                backward ref.
+     *
+     * @param fetcher The name of the fether to be used for this request.
```
```suggestion
     * @param fetcher The name of the fetcher to be used for this request.
```

luca-schlecker replied:
Already fixed.

mowi12 commented on src/main/kotlin/se/uulm/snowballr/backend/service/FetcherService.kt:
```diff
@@ -0,0 +1,134 @@
+
+    override suspend fun removeFetcher(name: String) {
+        if (fetchers.remove(name) != null) {
+            logger.info { "Successfully registered a fetcher: $name" }
```
I think this is not correct : )

luca-schlecker replied:
Already fixed.

mowi12 commented on src/main/kotlin/se/uulm/snowballr/backend/service/FetcherService.kt:
I'm not sure, so please correct me if I'm wrong. But this component will be used by different threads.

If yes, then your implementation should not use a simple `HashMap` but a `ConcurrentHashMap`.

As well as your `ensureFetcher` implementation. I think a better solution would be to do this:

```kotlin
private fun getFetcherOrThrow(name: String): IFetcher {
    return fetchers[name] ?: throw SnowballRException.FetcherException.UnknownFetcher(name)
}

override suspend fun searchPapers(fetcher: String, searchQuery: String, options: Map<String, String>): Set<Paper> {
    return getFetcherOrThrow(fetcher).searchPapers(searchQuery, options)
}
```

For example. This way we remove the code smell `!!` and ensure that we either have a fetcher or throw. In your code, the fetcher could be deleted after the `ensureFetcher` call finishes.

luca-schlecker replied:
I really like your suggestion! So much cleaner. Also: Very sharp of you to notice the concurrency issue. 🙏

mowi12 reviewed (changes requested):
I know ... I know ... I just noticed, that this is only about the call and not the whole fetcher implementation ...

You can either ignore all these comments or also work them in in the other issue which @Merseleo is reviewing.

None the less, maybe there are some new things, which are also relevant for the other issue which Merseleo did not find.

Since this is dependent on the other issue, how should we continue with this issue/PR?


And you did not add tests ... Please add them.

luca-schlecker reviewed (commented):
Thank you very much for the valuable feedback. I implemented it in #120 and rebased this branch. The code quality went way up ⏫.

luca-schlecker commented:
As #125 is merged now, I took the liberty and resolved all the issues you've pointed out with it here.
This PR has now shifted a bit due to the rename of #125's `FetcherService` to `FetcherManager`.

A new `FetcherService` is introduced more closely following the idea for them (i.e. directly returning API responses).

mowi12 commented on src/main/kotlin/se/uulm/snowballr/backend/service/FetcherService.kt:
If you look at the other service implementations you will see, that we add a comment to the concrete class instead of the interface to describe the service. This way `params` are also included such as different repos which are needed by service. Could you please change this and keep the style of the other services?

luca-schlecker replied:
Moved the comment over :+1:

mowi12 commented on src/main/kotlin/se/uulm/snowballr/backend/service/FetcherService.kt:
```diff
@@ -0,0 +1,34 @@
+     */
+    suspend fun getAvailableFetchers(): AvailableFetcherApis
+
+    // TODO: WIP. Future PRs will introduce more functionality.
```
I don't think this TODO is needed. Imo, we only need to mark something with TODO if there is no issue for this, or it's only a temporary fix.
But as it is clear, that this service implementation will grow there, we do not need to add a TODO.

mowi12 commented on src/main/kotlin/se/uulm/snowballr/backend/service/FetcherService.kt:
```diff
@@ -0,0 +1,34 @@
+ */
+interface IFetcherService {
+    /**
+     * Service implementation of [SnowballRService.getAvailableFetcherApis].
```
Please add the correct import for the `SnowballRService`.

mowi12 replied:
I closed this and added a new conversation with a suggestion.

mowi12 commented on src/main/kotlin/se/uulm/snowballr/backend/service/FetcherService.kt:
```diff
@@ -0,0 +1,34 @@
+import se.uulm.snowballr.backend.fetcher.FetcherManager
+import snowballr.Main.AvailableFetcherApis
+
+private val logger = KotlinLogging.logger {}
```
This logger is not needed. Are there other Issue in which you use this logger? If yes we can keep this, else I would remove this as it will be added when it will be used.

luca-schlecker replied:
Will remove it and reintroduce when really needed. :+1:

mowi12 commented on src/main/kotlin/se/uulm/snowballr/backend/service/FetcherService.kt:
```diff
@@ -0,0 +1,34 @@
+ * Default implementation of [IFetcherService].
+ */
+class FetcherService : IFetcherService {
+    private val fetcherManager = FetcherManager()
```
Why are you creating the fetcher manager like this?
Didn't you add the manager to the modules so it can be injected?

Have a look at the user service and how it uses the JWT service. I think this would be the correct way to do it.

You would also need to change the `MainService` file!

luca-schlecker replied:
Somehow forgot... Will change this.

mowi12 reviewed (changes requested):
Sorry, this took so long, I hadn't that much time the last days.

First of all, could you please add a comment to the `Module.kt` file so it mentions the creation of the `FetcherManager`?
Maybe also change the order so it gets created before the database? I'm not sure if the order even changes anything. But as the other Singletons such as the `PasswordUtils` are also instantiated at the start, I think this should also go there.
I just saw this, as I also looked at the other PR so I could understand this one.

Regarding this issue, there are some points you need to check out. I think we can improve the usage of the `FetcherManager` similar to how we used other managers/util classes.
The rest is only to be consistent.

And please add some tests, even if they are very simple. But we want to keep the coverage as high as possible : D

luca-schlecker commented:
> Sorry, this took so long, I hadn't that much time the last days.
> 
> First of all, could you please add a comment to the `Module.kt` file so it mentions the creation of the `FetcherManager`? Maybe also change the order so it gets created before the database? I'm not sure if the order even changes anything. But as the other Singletons such as the `PasswordUtils` are also instantiated at the start, I think this should also go there. I just saw this, as I also looked at the other PR so I could understand this one.
> 
> Regarding this issue, there are some points you need to check out. I think we can improve the usage of the `FetcherManager` similar to how we used other managers/util classes. The rest is only to be consistent.
> 
> And please add some tests, even if they are very simple. But we want to keep the coverage as high as possible : D

No worries, I was able to continue my work in different branches and will rebase accordingly.
I added the comments you suggested. The order should make no difference in this case, but I changed it anyway.
I responded to your comments and added tests. ^^

mowi12 commented on src/main/kotlin/se/uulm/snowballr/backend/service/FetcherService.kt:
```diff
@@ -0,0 +1,26 @@
+package se.uulm.snowballr.backend.service
+
+import se.uulm.snowballr.backend.fetcher.FetcherManager
+import snowballr.Main.AvailableFetcherApis
```
```suggestion
import snowballr.Main.AvailableFetcherApis
import se.uulm.snowballr.backend.grpc.SnowballRServer.SnowballRService
```

mowi12 commented on src/main/kotlin/se/uulm/snowballr/backend/service/FetcherService.kt:
```diff
@@ -0,0 +1,26 @@
+ * Orchestrates the paper fetching process. This service is responsible for
+ * retrieving papers for a specific request. It has to make sure that the
+ * correct fetchers (specified in the project settings) are accessed using the
+ * correct options (also specified in the project setitngs).
```
```suggestion
 * correct options (also specified in the project settings).
 *
 * @param fetcherManager The manager that handles the available fetchers and their configurations.
```

mowi12 commented on src/test/kotlin/se/uulm/snowballr/backend/service/FetcherServiceTest.kt:
Could you please move this file into a matching directory, similar to all the other service test files?
And also rename it to the Method your testing. Also similar to the service test files : D

mowi12 reviewed (changes requested):
Perfect! Almost done.

mowi12 commented on src/main/kotlin/se/uulm/snowballr/backend/service/FetcherService.kt:
```diff
@@ -0,0 +1,27 @@
+ * retrieving papers for a specific request. It has to make sure that the
+ * correct fetchers (specified in the project settings) are accessed using the
+ * correct options (also specified in the project settings).
+ */
```
```suggestion
 * correct options (also specified in the project settings).
 *
 * @param fetcherManager The [FetcherManager] that manages the available fetchers.
 */
```

mowi12 reviewed (changes requested):
I thought I wrote a comment regarding the last new conversation. Maybe I forgot it.
But this is the last change, then we are done!

luca-schlecker commented:
I implemented your suggestion, rebased the branch, and squashed the commits.

mowi12 reviewed (approved):
Perfect, lgtm!
